src/async_impl/client.rs

```diff
@@ -1971,7 +1994,16 @@ impl Client {
     ///
     /// This method fails whenever the supplied `Url` cannot be parsed.
     pub fn request<U: IntoUrl>(&self, method: Method, url: U) -> RequestBuilder {
-        let req = url.into_url().map(move |url| Request::new(method, url));
+        let maybe_url = match &self.inner.base_url {
+            Some(ref base_url) => Url::options()
+                .base_url(Some(base_url))
+                .parse(url.as_str())
+                .map_err(crate::error::builder)
+                .and_then(IntoUrlSealed::into_url),
```

Comment:
This is the only part of the code I find suspect, as maybe this should live in `IntoUrl`. I was not too sure on the best way to make changes there without changing the trait (that I was trying to avoid).

For the `base_url` behaviour, I have deferred to how it is implemented in the Url crate. This is documented here: https://docs.rs/url/latest/url/struct.Url.html#method.join